### PR TITLE
Refresh token public app support

### PR DESCRIPTION
**Pull request by sfcbetiuc**
- Added Refresh Token support for OAuth2 authentication
- Added Web/Public App support for OAuth2 authentication

### CODE REVIEW

File: src/test/java/com/exacttarget/fuelsdk/ETClientTest.java [hunk 1]
```diff
@@ -224,6 +195,129 @@ public void testSoapEndpointCaching()
...
+    public void createPayload_shouldReturnPayloadWithWebAppAttributes_whenApplicationTypeIsWeb(){
+        ETConfiguration configuration = client.getConfiguration();
+
+        configuration.set("applicationType", "web");
+        configuration.set("redirectURI", "redirectURI");
+        configuration.set("authorizationCode", "authorizationCode");
+
+        JsonObject payload = client.createPayload(configuration);
+
+        assertEquals(configuration.get("clientId"), payload.get("client_id").getAsString());
```
Comment (sfadincescu) [inline src/test/java/com/exacttarget/fuelsdk/ETClientTest.java]: too many asserts for one test. please use a data provider.

File: src/test/java/com/exacttarget/fuelsdk/ETClientTest.java [hunk 2]
```diff
@@ -224,6 +195,129 @@ public void testSoapEndpointCaching()
...
+
+    @Test
+    public void createPayload_shouldReturnPayloadWithServerAppAttributes_whenApplicationTypeIsServer(){
+        ETConfiguration configuration = client.getConfiguration();
+
+        configuration.set("applicationType", "server");
+
+        JsonObject payload = client.createPayload(configuration);
+
+        assertEquals(configuration.get("clientId"), payload.get("client_id").getAsString());
```
Comment (sfadincescu) [inline src/test/java/com/exacttarget/fuelsdk/ETClientTest.java]: same as above: too many asserts

File: src/test/java/com/exacttarget/fuelsdk/ETClientTest.java [hunk 3]
```diff
@@ -224,6 +195,129 @@ public void testSoapEndpointCaching()
...
+        assertEquals("client_credentials", payload.get("grant_type").getAsString());
+    }
+
+    @Test
+    public void createPayload_shouldReturnPayloadWithServerAppAttributes_whenApplicationTypeIsNotPassedInConfig(){
+        ETConfiguration configuration = client.getConfiguration();
+
+        JsonObject payload = client.createPayload(configuration);
+
+        assertEquals(configuration.get("clientId"), payload.get("client_id").getAsString());
```
Comment (sfadincescu) [inline src/test/java/com/exacttarget/fuelsdk/ETClientTest.java]: same here

File: src/main/java/com/exacttarget/fuelsdk/ETClient.java [hunk 4]
```diff
@@ -191,15 +204,25 @@ public ETClient(ETConfiguration configuration)
...
 
+    private void buildClients() throws ETSdkException {
+        authConnection = new ETRestConnection(this, authEndpoint, true);
+        requestToken();
+        restConnection = new ETRestConnection(this, endpoint);
+        fetchSoapEndpoint();
+        soapConnection = new ETSoapConnection(this, soapEndpoint, accessToken);
+    }
+
+    public static boolean isNullOrBlankOrEmpty(String applicationType) {
```
[the remark this answers is not in the document]
Comment (sfcbetiuc) [reply src/main/java/com/exacttarget/fuelsdk/ETClient.java]: Done.

File: src/main/java/com/exacttarget/fuelsdk/ETClient.java [hunk 5]
```diff
@@ -364,34 +363,58 @@ private synchronized String requestOAuth2Token()
...
+
         return accessToken;
     }
 
+    JsonObject createPayload(ETConfiguration configuration) {
+        JsonObject payload = new JsonObject();
+
+        payload.addProperty("client_id", configuration.get("clientId"));
+
+        if(configuration.get("applicationType").equals("web") || configuration.get("applicationType").equals("server")){
```
[the remark this answers is not in the document]
Comment (sfcbetiuc) [reply src/main/java/com/exacttarget/fuelsdk/ETClient.java]: Done.

**Pull request merged**